### PR TITLE
authz: simplify `PermsStore` logic with intarray

### LINKED ISSUE #11768: Simplify `PermsStore` logic with intarray extension
https://www.postgresql.org/docs/9.6/intarray.html

### PR DESCRIPTION
**Pull request by unknwon**
Uses operators and functions provided by Postgres' [intarray extension](https://www.postgresql.org/docs/9.6/intarray.html) to greatly simplify the logic of some `PermsStore` methods.

Regarding in-place changes in tests, I got confused in debugging SQLs with these IDs, so I choose to use mutually exclusive sets of IDs for users and repos.

**Please review by commit in split view**

Stack on https://github.com/sourcegraph/sourcegraph/pull/11457 and fixes #11768

### CODE REVIEW

**Comment by keegancsmith** [comment]
What is the cost to performance? From what I can tell an `intarray` is just a random access array, so any operation you do on it will now be `O(N)`. I suppose previously with the bitmaps you had to read back the whole structure first, which is `O(N)` (even though the subsequent manipulations would be `O(1)`)

File: enterprise/internal/db/perms_store.go [hunk 1]
```diff
@@ -1045,23 +1020,22 @@ VALUES
 ON CONFLICT ON CONSTRAINT
   repo_permissions_perm_unique
 DO UPDATE SET
-  user_ids = excluded.user_ids,
+  user_ids = UNIQ(SORT(repo_permissions.user_ids + excluded.user_ids)),
```
Comment (tsenart) [inline enterprise/internal/db/perms_store.go:860]: I think you can use `|` here and avoid the `UNIQ(SORT(`. Please validate.
```suggestion
  user_ids = repo_permissions.user_ids | excluded.user_ids,
```

File: enterprise/internal/db/perms_store.go [hunk 2]
```diff
@@ -400,24 +350,23 @@ VALUES
 ON CONFLICT ON CONSTRAINT
   user_permissions_perm_object_unique
 DO UPDATE SET
-  object_ids = excluded.object_ids,
+  object_ids = UNIQ(SORT(user_permissions.object_ids + excluded.object_ids)),
```
Comment (tsenart) [inline enterprise/internal/db/perms_store.go:353]: Same as my last comment, I think you can use `|`.

File: enterprise/internal/db/perms_store.go [hunk 3]
```diff
@@ -536,51 +539,39 @@ func (s *PermsStore) SetRepoPendingPermissions(ctx context.Context, accounts *ex
 	added := roaring.AndNot(p.UserIDs, oldIDs)
 	removed := roaring.AndNot(oldIDs, p.UserIDs)
 
-	// Load stored user IDs of both added and removed.
-	changedIDs := roaring.Or(added, removed).ToArray()
-
-	// In case there is nothing to add or remove.
-	if len(changedIDs) == 0 {
+	// In case nothing added or removed.
+	if added.GetCardinality() == 0 && removed.GetCardinality() == 0 {
```
Comment (tsenart) [inline enterprise/internal/db/perms_store.go:543]: Is there no method on this type like `IsEmpty()` or something?

File: enterprise/internal/db/perms_store.go [hunk 4]
```diff
@@ -757,26 +689,24 @@ VALUES
 ON CONFLICT ON CONSTRAINT
   user_pending_permissions_service_perm_object_unique
 DO UPDATE SET
-  object_ids = excluded.object_ids,
+  object_ids = UNIQ(SORT(user_pending_permissions.object_ids + excluded.object_ids)),
```
Comment (tsenart) [inline enterprise/internal/db/perms_store.go:692]: Ditto.

File: enterprise/internal/db/perms_store.go [hunk 5]
```diff
@@ -852,8 +827,11 @@ func (s *PermsStore) GrantPendingPermissions(ctx context.Context, userID int32,
 	}
 	up.IDs = roaring.Or(oldIDs, p.IDs)
 
-	up.UpdatedAt = txs.clock()
-	if q, err = upsertUserPermissionsBatchQuery(up); err != nil {
+	objectIDs := make([]int32, 0, up.IDs.GetCardinality())
+	for _, id := range up.IDs.ToArray() {
```
Comment (tsenart) [inline enterprise/internal/db/perms_store.go:831]: I don't think you need to call `ToArray` to iterate over this.

**Comment by keegancsmith** [comment]
@unknwon so i'm interested if we will end up in a place where we do auth filtering purely in SQL. IE if we only wanted a list of repo IDs from the DB we could just fetch that (authzFilter would be removed?). More concretely removing these lines and still respecting auth. https://github.com/sourcegraph/sourcegraph/blob/c6ca86f0b63f73a4f4ccd148f35049e45e2be3eb/internal/db/repos.go#L206-L210

**Comment by keegancsmith** [comment]
What is the status of this? If you would like a review, please add me back as a reviewer :)

**Comment by unknwon** [comment]
@keegancsmith Intend to resume the work here after https://github.com/sourcegraph/sourcegraph/pull/13708

**Comment by unknwon** [comment]
Close in favor of https://github.com/sourcegraph/sourcegraph/pull/16003.

**Pull request closed without merging**